laws/shared/src/main/scala/scalaz/laws/TypeclassLaws.scala

```diff
@@ -0,0 +1,217 @@
+    )
+  }
+}
+
+object MonadLaws {
+  @inline
+  def bindIdentity[F[_], A, T](in: F[A])(assert: (F[A], F[A]) => T)(implicit F: Monad[F]) =
+    assert(in, F.flatMap(in)(F.pure))
+}
+
+object CobindLaws {
+  @inline
+  def cobindAssoc[F[_], A, B, C, T](in: F[A])(fst: F[A] => B,
+                                              snd: F[B] => C)(assert: (F[C], F[C]) => T)(implicit F: Cobind[F]): T = {
+    import F.cobind
+    assert(
+      cobind(cobind(in)(fst))(snd),
+      cobind(in)((b: F[A]) => snd(cobind(b)(fst)))
+    )
+  }
+}
+
+object ComonadLaws {
+  @inline
+  def cobindIdentity[F[_], A, T](in: F[A])(assert: (F[A], F[A]) => T)(implicit F: Comonad[F]): T =
+    assert(in, F.cobind(in)(F.copoint))
+}
+
+object TraversableLaws {
+  // it's okay to fuse multiple `traverse` calls
+  // that can be in different applicatives
+  // into a single call over the composite applicative.
+  @inline
+  def traverseComposition[T[_]: Traversable, F[_]: Applicative, G[_]: Applicative, A, B, C, Test](
+    in: T[A]
+  )(f: A => F[B], g: B => G[C])(assert: (F[G[T[C]]], F[G[T[C]]]) => Test): Test =
+    assert(
+      in.traverse(f).map(_.traverse(g)),
+      Compose.run(in.traverse(a => Compose(f(a).map(g))))
+    )
+
+  @inline
+  def traverseIdentity[F[_], A, T](in: F[A])(assert: (F[A], F[A]) => T)(implicit F: Traversable[F]) =
+    assert(in, Identity.run(F.traverse(in)(Identity(_))))
+}
+
+object SemicategoryLaws {
+  @inline
+  def composeAssoc[F[_, _], A, B, C, D, T](fst: F[C, D], snd: F[B, C], thd: F[A, B])(
+    assert: (F[A, D], F[A, D]) => T
+  )(implicit F: Semicategory[F]): T = {
+    import F.compose
+    assert(
+      compose(compose(fst, snd), thd),
+      compose(fst, compose(snd, thd))
+    )
+  }
+}
+
+object CategoryLaws {
+  @inline
+  def composeRightIdentity[F[_, _], A, B, T](in: F[A, B])(assert: (F[A, B], F[A, B]) => T)(implicit F: Category[F]): T =
+    assert(in, F.compose(in, F.id))
+
+  @inline
+  def composeLeftIdentity[F[_, _], A, B, T](in: F[A, B])(assert: (F[A, B], F[A, B]) => T)(implicit F: Category[F]): T =
+    assert(in, F.compose(F.id, in))
+}
+
+object EqLaws {
+  // pass me two `A`'s
+  // and some way to observe an `A` as a `B`
+  // and I'll tell you if the observations agree
+  // with their equality.
+  @inline
+  def identity[A, B, T](fst: A, snd: A)(f: A => B)(assert: (Boolean, B, B) => T)(implicit A: Eq[A]): T =
+    assert(A.equal(fst, snd), f(fst), f(snd))
+
+  @inline
+  def reflexivity[A, T](in: A)(assert: Boolean => T)(implicit A: Eq[A]): T =
+    assert(A.equal(in, in))
+
+}
+
+object OrdLaws {
+
+  @inline
+  def antisymmetry[A, T](fst: A, snd: A)(assert: Boolean => T)(implicit A: Ord[A]): T = {
+    val mustBeEqual =
+      ((A.<=(fst, snd)) && (A.<=(snd, fst)))
+    val satisfiesAntisymmetry =
+      mustBeEqual == A.equal(fst, snd)
+    assert(satisfiesAntisymmetry)
+  }
+
+  @inline
+  def connex[A, T](fst: A, snd: A)(assert: Boolean => T)(implicit A: Ord[A]): T = {
```

Comment:
it feels like there should be a better `assert` type than this, but I don't think it's in testz yet, is it?